exp/addons/controllers/clusterresourceset_helpers.go

```diff
@@ -0,0 +1,160 @@
+		}
+
+		clusterResourceSetBinding.Name = cluster.Name
+		clusterResourceSetBinding.Namespace = cluster.Namespace
+		clusterResourceSetBinding.OwnerReferences = util.EnsureOwnerRef(clusterResourceSetBinding.OwnerReferences, metav1.OwnerReference{
+			APIVersion: clusterv1.GroupVersion.String(),
+			Kind:       "Cluster",
+			Name:       cluster.Name,
+			UID:        cluster.UID,
+		})
+		clusterResourceSetBinding.OwnerReferences = util.EnsureOwnerRef(clusterResourceSetBinding.OwnerReferences, *metav1.NewControllerRef(clusterResourceSet, clusterResourceSet.GroupVersionKind()))
+
+		if err := r.Client.Create(ctx, clusterResourceSetBinding); err != nil {
+			if apierrors.IsAlreadyExists(err) {
+				if err = r.Client.Get(ctx, clusterResourceSetBindingKey, clusterResourceSetBinding); err != nil {
+					return nil, err
+				}
+				return clusterResourceSetBinding, nil
+			}
+			return nil, errors.Wrapf(err, "failed to create clusterResourceSetBinding for cluster: %s/%s", cluster.Namespace, cluster.Name)
+		}
+	}
+
+	return clusterResourceSetBinding, nil
+}
+
+// initClusterResourceSetBinding initializes ClusterResourceSetBinding object for a ClusterResourceSet if not initialized already.
+// clusterResourceSetBinding should not be nil
+func initClusterResourceSetBinding(clusterResourceSetBinding *addonsv1.ClusterResourceSetBinding, clusterResourceSet *addonsv1.ClusterResourceSet) {
+	if clusterResourceSetBinding.Spec.Bindings == nil {
+		clusterResourceSetBinding.Spec.Bindings = map[string]addonsv1.ResourcesSetBinding{}
+	}
+
+	// Add ClusterResourceSet to clusterResourceSetBinding if not exist
+	if _, ok := clusterResourceSetBinding.Spec.Bindings[clusterResourceSet.Name]; !ok {
+		clusterResourceSetBinding.Spec.Bindings[clusterResourceSet.Name] = addonsv1.ResourcesSetBinding{Resources: map[string]addonsv1.ResourceBinding{}}
+	}
+}
+
+// getConfigMap retrieves any ConfigMap from the given name and namespace.
+func getConfigMap(ctx context.Context, c client.Client, configmapName types.NamespacedName) (*corev1.ConfigMap, error) {
+	configMap := &corev1.ConfigMap{}
+	configMapKey := client.ObjectKey{
+		Namespace: configmapName.Namespace,
+		Name:      configmapName.Name,
+	}
+	if err := c.Get(ctx, configMapKey, configMap); err != nil {
+		return nil, err
+	}
+
+	return configMap, nil
+}
+
+// getSecret retrieves any Secret from the given secret name and namespace.
+func getSecret(ctx context.Context, c client.Client, secretName types.NamespacedName) (*corev1.Secret, error) {
+	secret := &corev1.Secret{}
+	secretKey := client.ObjectKey{
+		Namespace: secretName.Namespace,
+		Name:      secretName.Name,
+	}
+	if err := c.Get(ctx, secretKey, secret); err != nil {
+		return nil, err
+	}
+
+	return secret, nil
+}
+
+func computeHash(dataArr [][]byte) string {
+	hash := sha256.New()
+	for i := range dataArr {
+		_, err := hash.Write(dataArr[i])
+		if err != nil {
+			continue
+		}
+	}
+	return base64.URLEncoding.EncodeToString(hash.Sum(nil))
```

Comment:
Any reason to use base64 over hex? (`fmt.Printf("%x", hash.Sum(nil))`)?

Reply:
No reason, some are using base64 in k/k.
Changed to base16.